### DISCUSSION TITLE
哪个更好？Xray前端还是Caddy前端？

### DISCUSSION THREAD

**Discussion by houmie** [lxhao61/integrated-examples#165] [Q&A] [answered]
您好，大师，

我非常喜欢您制作的精细示例，这两个是我最喜欢的：
Xray(M+H+K+G+B+A)+Caddy(N+T)是Xray前置。
Xray(M+F+H+K+G+B+A)+Caddy(N)是使用4层转发的Caddy前置。

1）在隐藏和安全性方面，哪一个更好？因为它们都有诱饵网站（分流机制）。你会选择哪一个？

2）在第一个示例中也支持Trojan+TCP+TLS (F)。我认为主要的区别是第一个示例支持Trojan-Go，但第二个示例不支持。

3）您对gRPC有什么看法？
Trojan+gRPC+TLS  (+CDN)
VLESS+gRPC+REALITY 
VLESS+gRPC+TLS (+CDN)

谢谢，

**Comment by houmie** [comment]
啊，我发现了另一个区别。Xray(M+F+H+G+B+A)+Caddy(N)的示例不支持Trojan-WS-TLS。我在想为什么...

**Comment by lxhao61** [comment, marked as answer]
1）、在隐藏和安全性方面一样，无区别。我选第一个示例，主要是其配置相对简单、且对Trojan性能不敏感。
2）、两者主要区别在Trojan：第一个示例是由Caddy的caddy-trojan插件实现，而第二个示例是由Xray配置实现。第二个示例加上Trojan+WebSocket+TLS应用也可支持Trojan-Go，且其WebSocket应用的性能及兼容性更好；默认不带Trojan+WebSocket+TLS应用在于已有升级的HTTPUpgrade应用，重复无必要。
3）、VLESS+gRPC+REALITY有TLS in TLS风险，不要使用“偷自己的证书”配置。Trojan+gRPC+TLS (+CDN)与VLESS+gRPC+TLS (+CDN)除了使用密码与使用UUID外，无差异。

**Comment by houmie** [comment]
谢谢您，师傅。

我还有一些问题想要请教，因为我并没有找到确切的答案。我希望您能帮助我。

1）为什么VMess+WebSocket+TLS仍然如此受欢迎？在2020年，有报告称VMess已被GFW破解. https://gfw.report/blog/v2ray_weaknesses/en/
为什么不使用VLESS+WebSocket+TLS。那不是更安全吗？

2）如您所建议的，我正在考虑在我的VPS上使用Xray(M+H+D+G+B+A)+Caddy(N+T)的示例。VLESS+Reality的变体非常安全，但是如果有人使用VLESS+mKCP+seed并被GFW抓住，那么整个服务器的IP就会被列入黑名单，对吧？然后VLESS+Reality的解决方案也将无法工作，因为整个服务器已经被列入黑名单。那么，是否更好的做法是，为VLESS+Reality的解决方案设置一个单独的VPS，然后为像VLESS+mKCP+seed这样的不太安全的协议设置另一个VPS？

我是指像这样的配置：
VPS 1 -> Xray(M+K)+Caddy，以及 
VPS 2 -> Xray(F+H+G+B+A)+Caddy。

3）哪种协议最适合游戏？我听说VLESS+mKCP+seed是游戏、VOIP和流媒体的绝佳选择，这是真的吗？

4）VLESS+Vision+REALITY和VLESS+HTTP+REALITY之间的实际区别是什么？第一个由于没有TLS-in-TLS而更安全，而第二个由于H2而更快？

感谢您向我解释这些问题。

**Comment by lxhao61** [comment]
1）、见如下截图：
![image](https://github.com/lxhao61/integrated-examples/assets/24571146/270f15de-c9dd-4ab9-a20f-749c0f25f142)
2）、见如下截图：
![image](https://github.com/lxhao61/integrated-examples/assets/24571146/ae03bb5f-57f8-41d1-952e-32fc8601d962)
![image](https://github.com/lxhao61/integrated-examples/assets/24571146/fb53e30b-67c7-4b1d-b80f-5f15ddccd520)
3）、是的。
4）、是的。

**Reply by houmie** [reply]
1) 您的截图对我来说非常难以翻译成英文，我并没有完全理解。特别是这部分：

否则推荐使用自身不带加空的VLESS或Trojan办议配置。

所以，您的意思是，如果CDN不值得信任，那么最好使用VLESS+HTTPUpgrade+TLS。
但如果CDN值得信任，如Cloudflare，那么使用VMESS+HTTPUpgrade+TLS是安全的。
但为什么要冒险呢？为什么不始终使用VLESS+HTTPUpgrade+TLS而非VMESS+HTTPUpgrade+TLS？

2) 我认为我理解您的意思。您同意我对安全性的担忧。所以，您认为从Xray(M+H+D+G+B+A)+Caddy(N+T)中删除mKCP、HTTPUpgrade、HTTP和gRPC传输方法，并只允许VLESS+VISION+TCP+REALITY，会更好。

3) 最后一个问题，我听说gRPC非常快，但易于被检测。一个朋友告诉我，VLESS+gRPC+TLS，Mux设置为4或8，速度极快。您会推荐这种组合吗？

谢谢

**Comment by lxhao61** [comment]
晕！仔细看项目介绍及注意事项、备注等。
1、理解反了！截图来自项目首页注意事项7与8。
2、截图来自注意事项7。另外mKCP、Websocket、H2C/gRPC配置可以是用备用，不一定要必须使用呀。若你一定要所谓绝对[安全](https://github.com/lxhao61/integrated-examples/discussions/152)，那直接单一配置即可呀。
3、gPRC传输方式都一样，可随便选择：反正我提供的此类示例都支持非AES算法密码套件配置。我推荐Shadowsocks+gRPC+TLS，在于支持应用多。

**Comment by houmie** [comment]
哇，截图来自哪里并不重要。我向你明确地提问了关于VMESS和VLESS的问题，你却不能给我一个直接的答案，反而转移到与我的问题无关的注意事项上。哈哈

**Comment by lxhao61** [comment]
仔细看项目介绍及注意事项、备注等！这些早就清清楚楚写出来告知了，自己不仔细看反而要求别人再次一一给解惑，太搞笑了吧？！
